linenoise/linenoise/linenoise.h

```diff
@@ -0,0 +1,335 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __LINENOISE_H
+#define __LINENOISE_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include "esp_err.h"
+
+/*
+ * IMPORTANT: This library is not thread safe.
+ */
+
+/*------------------------------------------------------------------------------------------------
+ *NEW LINENOISE API
+ *------------------------------------------------------------------------------------------------*/
+
+/**
+ * @brief Stores a list of completion strings.
+ */
+typedef struct linenoise_completions {
+  size_t len;   /**< Number of completions. */
+  char **cvec;  /**< Array of completion strings. */
+} linenoise_completions_t;
+
+/**
+ * @brief Callback function for providing completions.
+ *
+ * @param str Input string from the user.
+ * @param lc Completions object to be populated.
+ */
+typedef void (*linenoise_completion_callback)(const char *str, linenoise_completions_t *lc);
+
+/**
+ * @brief Callback function for providing hints during input.
+ *
+ * @param str Input string.
+ * @param color Pointer to set the color of the hint text.
+ * @param bold Pointer to set the bold attribute (non-zero for bold).
+ * @return A dynamically allocated hint string or NULL if no hint is available.
+ */
+typedef char* (*linenoise_hints_callback)(const char *str, int *color, int *bold);
+
+/**
+ * @brief Callback function to free memory returned by hints callback.
+ *
+ * @param ptr Pointer to the hint string to be freed.
+ */
+typedef void (*linenoise_free_hints_callback)(void *ptr);
+
+/**
+ * @brief Function pointer type for reading bytes.
+ *
+ * @param fd File descriptor.
+ * @param buf Buffer to store the read bytes.
+ * @param count Number of bytes to read.
+ * @return Number of bytes read, or -1 on error.
+ */
+typedef ssize_t (*linenoise_read_bytes_fn)(int fd, void* buf, size_t count);
+
+/**
+ * @brief Function pointer type for writing bytes.
+ *
+ * @param fd File descriptor.
+ * @param buf Buffer containing bytes to write.
+ * @param count Number of bytes to write.
+ * @return Number of bytes written, or -1 on error.
+ */
+typedef ssize_t (*linenoise_write_bytes_fn)(int fd, const void* buf, size_t count);
+
+/**
+ * @brief Structure defining the parameters needed by a linenoise
+ * instance
+ */
+typedef struct linenoise_instance_param {
+    const char *prompt;
+    size_t max_cmd_line_length;
+    int history_max_length;
+    int in_fd;
+    int out_fd;
+    bool allow_multi_line;
+    bool allow_empty_line;
+    bool allow_dumb_mode;
+    linenoise_completion_callback completion_cb;
+    linenoise_hints_callback hints_cb;
+    linenoise_free_hints_callback free_hints_cb;
+    linenoise_read_bytes_fn read_bytes_fn;
+    linenoise_write_bytes_fn write_bytes_fn;
+    char **history;
+} linenoise_instance_param_t;
```

Comment:
For consistency with other components, let's call this "config"

```suggestion
} linenoise_instance_config_t;
```